share/availability/full/reconstruction_test.go

```diff
@@ -79,79 +118,119 @@ func TestShareAvailable_OneFullNode(t *testing.T) {
+
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+			defer cancel()
+
+			net := availability_test.NewTestDAGNet(ctx, t)
+			source, root := RandNode(net, origSquareSize, tt.recoverability)
+
+			// create two full nodes and ensure they are disconnected
+			full1 := Node(net)
+			full2 := Node(net)
+
+			// pre-connect fulls
+			net.Connect(full1.ID(), full2.ID())
+			// ensure fulls and source are not connected
+			// so that fulls take data from light nodes only
+			net.Disconnect(full1.ID(), source.ID())
+			net.Disconnect(full2.ID(), source.ID())
+
+			// start reconstruction for fulls
+			errg, errCtx := errgroup.WithContext(ctx)
+			errg.Go(func() error {
+				return full1.SharesAvailable(errCtx, root)
+			})
+			errg.Go(func() error {
+				return full2.SharesAvailable(errCtx, root)
+			})
+
+			// create light nodes and start sampling for them immediately
+			lights1, lights2 := make(
+				[]*availability_test.TestNode, lightNodes/2),
+				make([]*availability_test.TestNode, lightNodes/2)
+			for i := 0; i < len(lights1); i++ {
+				lights1[i] = light.Node(net)
+				go func(i int) {
+					err := lights1[i].SharesAvailable(ctx, root)
+					if err != nil {
+						t.Log("light1 errors:", err)
+					}
+				}(i)
+
+				lights2[i] = light.Node(net)
+				go func(i int) {
+					err := lights2[i].SharesAvailable(ctx, root)
+					if err != nil {
+						t.Log("light2 errors:", err)
+					}
+				}(i)
 			}
-		}(i)
 
-		lights2[i] = light.Node(net)
-		go func(i int) {
-			err := lights2[i].SharesAvailable(ctx, root)
-			if err != nil {
-				t.Log("light2 errors:", err)
+			// shape topology
+			for i := 0; i < len(lights1); i++ {
+				// ensure lights1 are only connected to full1
+				net.Connect(lights1[i].ID(), full1.ID())
+				net.Disconnect(lights1[i].ID(), full2.ID())
+				// ensure lights2 are only connected to full2
+				net.Connect(lights2[i].ID(), full2.ID())
+				net.Disconnect(lights2[i].ID(), full1.ID())
 			}
-		}(i)
-	}
 
-	// shape topology
-	for i := 0; i < len(lights1); i++ {
-		// ensure lights1 are only connected to full1
-		net.Connect(lights1[i].ID(), full1.ID())
-		net.Disconnect(lights1[i].ID(), full2.ID())
-		// ensure lights2 are only connected to full2
-		net.Connect(lights2[i].ID(), full2.ID())
-		net.Disconnect(lights2[i].ID(), full1.ID())
-	}
+			// start connection lights with sources
+			for i := 0; i < len(lights1); i++ {
+				net.Connect(lights1[i].ID(), source.ID())
+				net.Connect(lights2[i].ID(), source.ID())
+			}
 
-	// start connection lights with sources
-	for i := 0; i < len(lights1); i++ {
-		net.Connect(lights1[i].ID(), source.ID())
-		net.Connect(lights2[i].ID(), source.ID())
+			err := errg.Wait()
+			if tt.expectedFailure {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+		})
 	}
-
-	err := errg.Wait()
-	require.NoError(t, err)
 }
 
 // TestShareAvailable_DisconnectedFullNodes asserts that two disconnected full
```

Comment:
Random thought: Just curious, if it would still work same for similar topology, but for more than 2 full nodes.

Reply:
It should, yes